[lldb/Target] Delay image loading after corefile process creation

medismailben opened the pull request:
This patch is a follow-up to db223b7f01f7. Similarly to it, it changes the timing of binary image loading for the ProcessMachCore plugin.

This issue came up after getting reports of scripting resources that would fail to execute because they relied on data provided by the corefile process (i.e. for reading memory). However, rior to this change, the scripting resource loading would happen as part of the binary image loading, which in turns happened before the process finished being created.

This patch address that issue by delaying the binary image loading phase until we receive the corefile process stop event event, ensuring that the process is fully formed.

jasonmolenda reviewed (approved):
Yes this looks good to me.  I originally had a little concern about changing the ordering of how we call the methods in the corefile process plugins, but I don't think that's going to be an issue for them.

jimingham commented:
Is there a reason this is DidLoadCore only has a non-trivial implementation for Mach?

The generic and Mach parts of this are fine, so if there's a reason why this isn't needed for the other platform's core files, then this LGTM as well.

medismailben committed:
[lldb/Target] Delay image loading after corefile process creation

This patch is a follow-up to db223b7f01f7. Similarly to it, it changes
the timing of binary image loading for the ProcessMachCore plugin.

This issue came up after getting reports of scripting resources that
would fail to execute because they relied on data provided by the corefile
process (i.e. for reading memory). However, rior to this change, the
scripting resource loading would happen as part of the binary image
loading, which in turns happened before the process finished being created.

This patch address that issue by delaying the binary image loading phase
until we receive the corefile process stop event event, ensuring that the
process is fully formed.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

medismailben commented:
Add test like @jimingham suggested.

clayborg commented on lldb/include/lldb/Target/Process.h:
```diff
@@ -614,6 +614,8 @@ class Process : public std::enable_shared_from_this<Process>,
     return error;
   }
 
+  virtual void DidLoadCore() {}
```
We could avoid adding the DidLoadCore and re-use the existing Process DidAttach:
```
class Process {
...
  virtual void DidAttach(ArchSpec &process_arch) { process_arch.Clear(); }
```

medismailben replied:
Good point! At least for core file process plugins, the `DidAttach` override shouldn't be implemented so we should be good.

clayborg replied:
Yeah, I checked and none of the core file plug-ins currently override the DidAttach() (mach-o, ELF or minidump).

JDevlieghere replied:
Would it be worth to rename that function to `DidAttachOrLoadCore` (or just `DidAttachOrLoad`) to make it obvious that it handles both cases? I'm all for reusing `DidAttach` but it does seem a bit weird for a core file process.

clayborg replied:
I think it makes sense in the process plug-in landscape. We are attaching to a core file image from a process plug-in perspective. I would rather add a `DidLoadCore()` over renaming to `DidAttachOrLoadCore()` if we do need to make a change though.

medismailben replied:
I agree with @clayborg: attaching to a process and loading a core file are mutually exclusive actions so I can use re-using `DidAttach` in core file process plugins. However, if we rename `DidAttach` to `DidAttachOrLoadCore` then that would be called for non-corefile process plugins and I think that would be confusing to that a "LoadCore" mention in the context of the regular process.

@JDevlieghere if you think that using `DidAttach` in the core file process plugin context is confusing, I'm fine going back to my original implementation using `DidLoadCore`.

JDevlieghere replied:
Okay, so if semantically this doesn't make sense (which is what I'm hearing) then I don't why we would reuse `DidAttach` (besides it just happens to be there and unused in the core case) and we should have a `DidLoadCore` as per the current patch.

clayborg commented on lldb/source/Target/Process.cpp:
```diff
@@ -2668,7 +2655,23 @@ Status Process::LoadCore() {
       error.SetErrorString(
           "Did not get stopped event after loading the core file.");
+    } else {
+      DidLoadCore();
```
Call `DidAttach(...)` instead?

clayborg reviewed (commented):
Looks good, just a question if we want to use the existing Process::DidAttach() instead of adding a new Process::DidLoadCore()

JDevlieghere reviewed (approved):
LGTM per my last comment.

medismailben commented:
Reverting because this broke some tests:
```
********************
Failed Tests (5):
  lldb-api :: functionalities/postmortem/elf-core/TestLinuxCore.py
  lldb-api :: functionalities/postmortem/mach-core/TestMachCore.py
  lldb-api :: functionalities/postmortem/netbsd-core/TestNetBSDCore.py
  lldb-api :: functionalities/unwind/noreturn/module-end/TestNoReturnModuleEnd.py
  lldb-api :: tools/lldb-dap/coreFile/TestDAP_coreFile.py
  ```
